microsoft/utils/setup_aitl.py

```diff
@@ -0,0 +1,297 @@
+        return False
+
+    role_previous = previous_roles[0]
+    role_current = current_roles[0]
+
+    is_actions_match = _compare_str_list(
+        role_previous["permissions"][0]["actions"],
+        role_current["permissions"][0]["actions"],
+    )
+    is_dataactions_match = _compare_str_list(
+        role_previous["permissions"][0]["dataActions"],
+        role_current["permissions"][0]["dataActions"],
+    )
+    is_scopes_match = _compare_str_list(
+        role_previous["assignableScopes"], role_current["assignableScopes"]
+    )
+
+    if (
+        (role_previous["description"] == role_current["description"])
+        and (role_previous["roleName"] == role_current["roleName"])
+        and is_actions_match
+        and is_dataactions_match
+        and is_scopes_match
+    ):
+        return True
+    else:
+        return False
+
+
+def _check_same_role_existed(exist_role_json: str, target_role_json: str) -> bool:
+    # exist_role_json from az role list output, is an array: [role...]
+    # target_role_json from cmdline input, format is different from exist_role_json
+    if exist_role_json == "" or exist_role_json == "":
+        return False
+
+    exist_role_array = json.loads(exist_role_json)
+    if len(exist_role_array) != 1:
+        return False
+    exist_role = exist_role_array[0]
+
+    target_role = json.loads(target_role_json)
+
+    is_actions_match = _compare_str_list(
+        exist_role["permissions"][0]["actions"], target_role["Actions"]
+    )
+    is_dataactions_match = _compare_str_list(
+        exist_role["permissions"][0]["dataActions"], target_role["DataActions"]
+    )
+    is_scopes_match = _compare_str_list(
+        exist_role["assignableScopes"], target_role["AssignableScopes"]
+    )
+
+    if (
+        (exist_role["description"] == target_role["Description"])
+        and (exist_role["roleName"] == target_role["Name"])
+        and is_actions_match
+        and is_dataactions_match
+        and is_scopes_match
+    ):
+        return True
+    else:
+        return False
+
+
+def _wait_role_propagate(
+    role_name: str, subscription_id: str, role_before_update: str = ""
+) -> None:
+    print("In _wait_role_propagate: waiting for the role changes to propagate")
+    start_time = time.time()
+
+    change_count = 0
+    count = 0
+    max_count = 30
+    # check until it's the same for multiple times to make sure cache is refreshed.
+    while change_count < 24:
+        if count > max_count:
+            print(f"Fail to detect changed custom role after {count} times")
+            sys.exit(1)
+
+        print("Checking change applied...")
+
+        get_role_cmd_result = _call_cmd(
+            f'az role definition list --name "{role_name}" --scope "{subscription_id}'
+        )
+        role_after_update = get_role_cmd_result.stdout
+        if role_after_update.startswith("[]"):  # for role creation
+            time.sleep(5)
```

Comment:
It should be like below and not need to sleep and continue.
```python
        if not role_after_update.startswith("[]"):
```

Reply:
I don't think this is a good idea. After the creation cmd sent, it takes about 1~2 min to finish the role creation. We need to sleep for waiting. If not sleep, the program will continue checking role all the time before it's finished creation. I am not sure  this is what we want.

Reply:
It's not to change the logic, the full change is like below,

```python
        if not role_after_update.startswith("[]"): 
            is_roles_match = _compare_role_setting(role_before_update, role_after_update)
            if is_roles_match is False:
                print("Found changed results.")
                count += 1
                change_count += 1
            else:
                change_count = 0

        time.sleep(5)
```

Reply:
I changed to "not role_after_update.startswith("[]"):", but still waiting